src/com/unbounce/treajure/concurrent/mvar.clj

```diff
@@ -0,0 +1,82 @@
+(ns com.unbounce.treajure.concurrent.mvar
+  "Haskell's MVar abstraction in Clojure.
+
+  Check: http://chimera.labs.oreilly.com/books/1230000000929/ch07.html for more
+  info around MVars."
+  (:import [java.util.concurrent Semaphore]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Lightweight Performance mutable tuple implementation in Clojure
+;; (Look ma! no atoms!)
+
+(defprotocol IMCons
+  (mcar [p])
+  (mcdr [p])
+  (set-mcar! [p val])
+  (set-mcdr! [p val]))
+
+(deftype MCons [^{:volatile-mutable true} car
+                ^{:volatile-mutable true} cdr]
+  IMCons
+  (mcar [this] car)
+  (mcdr [this] cdr)
+  (set-mcar! [this val]
+    (set! car val))
+  (set-mcdr! [this val]
+    (set! cdr val)))
+
+(defn mcons [a b]
+  (MCons. a b))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; MVar abstraction
+
+(declare empty-mvar mvar put-mvar take-mvar read-mvar)
+
+(defn empty-mvar
+  "Create an empty MVar"
+  []
+  (let [take-sem (Semaphore. 1)
+        put-sem  (Semaphore. 1)]
+    (.acquire take-sem)
```

Comment:
Blocking operations with no time out, like this one, really freaks me out. If there's one thing I've learned from Erlang is: never block for ever, always have a timeout (and if not timeout is specifically defined, use a default short timeout - Erlang uses 3 seconds).

Whatever you end up using to back the mvar (semaphores, queues...), please make sure nothing has the potential to block for ever.